src/ucp/tag/rndv.c

```diff
@@ -497,37 +543,9 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),
         length = ucs_min(chunk, rndv_req->send.length - offset);
     }
 
-    /* ensure that the current length is over min_zcopy */
-    length = ucs_max(length, min_zcopy);
-
-    /* ensure that tail (rest of message) is over min_zcopy */
-    ucs_assertv(rndv_req->send.length >= (offset + length),
-                "send_length=%zu, offset=%zu, length=%zu",
-                rndv_req->send.length, offset, length);
-    tail = rndv_req->send.length - (offset + length);
-    if (ucs_unlikely((tail != 0) && (tail < min_zcopy))) {
-        /* ok, tail is less get zcopy minimal & could not be processed as
-         * standalone operation */
-        /* check if we have room to increase current part and not
-         * step over max_zcopy */
-        if (length < (max_zcopy - tail)) {
-            /* if we can increase length by min_zcopy - let's do it to
-             * avoid small tail (we have limitation on minimal get zcopy) */
-            length += tail;
-        } else {
-            /* reduce current length by align or min_zcopy value
-             * to process it on next round */
-            ucs_assert(length > ucs_max(min_zcopy, align));
-            length -= ucs_max(min_zcopy, align);
-        }
-    }
-
-    ucs_assertv(length >= min_zcopy, "length=%zu, min_zcopy=%zu",
-                length, min_zcopy);
-    ucs_assertv(((rndv_req->send.length - (offset + length)) == 0) ||
-                ((rndv_req->send.length - (offset + length)) >= min_zcopy),
-                "send_length=%zu, offset=%zu, length=%zu, min_zcopy=%zu",
-                rndv_req->send.length, offset, length, min_zcopy);
+    length = ucp_rndv_adjust_zcopy_length(min_zcopy, max_zcopy, align,
```

Comment:
remove line 543

Reply:
it could't be removed, since it calculates the length of the current chunk - i.e. minimum between current chunk's length and remaining length to be sent

Reply:
so, maybe in line 543 do `length = rndv_req->send.length - offset`

Reply:
then we should use `chunk` instead of `max_zcopy`:
```
diff --git a/src/ucp/tag/rndv.c b/src/ucp/tag/rndv.c
index 5ebe0f7..c02f5e9 100644
--- a/src/ucp/tag/rndv.c
+++ b/src/ucp/tag/rndv.c
@@ -535,15 +535,16 @@ UCS_PROFILE_FUNC(ucs_status_t, ucp_rndv_progress_rma_get_zcopy, (self),

     if ((offset == 0) && (remaining > 0) && (rndv_req->send.length > ucp_mtu)) {
         length = ucp_mtu - remaining;
+        chunk  = max_zcopy;
     } else {
         chunk = ucs_align_up((size_t)(ucs_min(rndv_req->send.length /
                                               rndv_req->send.rndv_get.lane_count,
                                               max_zcopy) * config->tag.rndv.scale[lane]),
                              align);
-        length = ucs_min(chunk, rndv_req->send.length - offset);
+        length = rndv_req->send.length - offset;
     }

-    length = ucp_rndv_adjust_zcopy_length(min_zcopy, max_zcopy, align,
+    length = ucp_rndv_adjust_zcopy_length(min_zcopy, chunk, align,
                                           rndv_req->send.length, offset,
                                           length);
```
is it ok?

Reply:
yeah, only question: why changed `ucp_mtu - remaining` to `max_zcopy` ?

Reply:
> why changed `ucp_mtu - remaining` to `max_zcopy` ?

it was not changed, just need to initialize `chunk` to `max_zcopy` to pass `chunk` instead of `max_zcopy` in `ucp_rndv_adjust_zcopy_length`

Reply:
> it was not changed, just need to initialize `chunk` to `max_zcopy` to pass `chunk` instead of `max_zcopy` in `ucp_rndv_adjust_zcopy_length`

right, so ok

Reply:
> > it was not changed, just need to initialize `chunk` to `max_zcopy` to pass `chunk` instead of `max_zcopy` in `ucp_rndv_adjust_zcopy_length`
> 
> right, so ok

done :)